### PR TITLE
EE-1185: add TrackingCopy::get_keys

### PR DESCRIPTION
**Pull request by henrytill**
This PR adds `get_keys` method to TrackingCopy which reads through the cache to `StateReader::keys_with_prefix`.

The caching behavior is marginally different than how we handle key-value pairs: cached writes must be unioned with reads rather than replace them.

### CODE REVIEW

File: types/src/key.rs [hunk 1]
```diff
@@ -350,37 +377,30 @@ impl From<ContractPackageHash> for Key {
 impl ToBytes for Key {
     fn to_bytes(&self) -> Result<Vec<u8>, Error> {
         let mut result = bytesrepr::unchecked_allocate_buffer(self);
+        result.push(self.tag());
```
Comment (xcthulhu) [inline types/src/key.rs:380]: From CI:

```
test storage::trie::tests::proptests::bincode_roundtrip_key ... ok
```

:heavy_check_mark:

File: execution_engine/src/core/tracking_copy/mod.rs [hunk 2]
```diff
@@ -231,6 +266,29 @@ impl<R: StateReader<Key, StoredValue>> TrackingCopy<R> {
...
+            None => {
+                let key_tag = key_tag.to_owned();
+                let keys = self
+                    .reader
+                    .keys_with_prefix(correlation_id, &[key_tag as u8])?;
+                ret.extend(keys);
+                self.cache.insert_key_tag_read(key_tag, ret.to_owned())
+            }
+        }
+        if let Some(keys) = self.cache.get_key_tag_muts_cached(&key_tag) {
```
Comment (xcthulhu) [inline execution_engine/src/core/tracking_copy/mod.rs:286]: I am a bit confused here should `key_tag_reads_cached` and `key_tag_muts_cached` be disjoint?  Or is the idea that `key_tag_muts_cached` masks `key_tag_reads_cached` and they can have duplicates?

Comment (xcthulhu) [reply execution_engine/src/core/tracking_copy/mod.rs:286]: I see now; that masking happens elsewhere.  This is for key enumeration; if you mutated a key these sets will contain duplicates.

File: execution_engine/src/core/tracking_copy/mod.rs [hunk 3]
```diff
@@ -231,6 +266,29 @@ impl<R: StateReader<Key, StoredValue>> TrackingCopy<R> {
...
+            Some(keys) => ret.extend(keys),
+            None => {
+                let key_tag = key_tag.to_owned();
+                let keys = self
+                    .reader
+                    .keys_with_prefix(correlation_id, &[key_tag as u8])?;
+                ret.extend(keys);
+                self.cache.insert_key_tag_read(key_tag, ret.to_owned())
+            }
+        }
```
Comment (xcthulhu) [inline execution_engine/src/core/tracking_copy/mod.rs:285]: You could do something like:

```rust
        // Do a cached read of the keys under a tag
        let key_tag_reads = match self.cache.get_key_tag_reads_cached(&key_tag) {
            Some(keys) => keys,
            None => {
                let key_tag = key_tag.to_owned();
                let keys = self
                    .reader
                    .keys_with_prefix(correlation_id, &[key_tag as u8])?;
                self.cache.insert_key_tag_read(key_tag, keys.to_owned())
                keys
            }
        }
        ret.extend(key_tag_reads);
```

**Comment by xcthulhu** [comment]
LGTM

**Review by xcthulhu — APPROVED** [review]

File: execution_engine/src/core/tracking_copy/mod.rs (hunk 2 above)
Comment (henrytill) [reply execution_engine/src/core/tracking_copy/mod.rs:286]: yah

File: execution_engine/src/core/tracking_copy/mod.rs (hunk 3 above)
Comment (henrytill) [reply execution_engine/src/core/tracking_copy/mod.rs:285]: Good one, will do.

Comment (henrytill) [reply execution_engine/src/core/tracking_copy/mod.rs:285]: Actually some impediments:
- currently `StateReader::keys_with_prefix` returns a `Vec` of keys.  I could change this to a `BTreeSet`, but...
- there's an ownership problem: `TrackingCopyCache::get_key_tags_reads_cached` returns a reference to the set but `StateReader::keys_with_prefix` returns the set itself.  The latter is unavoidable, but I don't want to have to clone if we change the former.

I may still change `StateReader::keys_with_prefix` to return a `BTreeSet` for consistency, but will likely do in a follow-up PR.

Comment (xcthulhu) [reply execution_engine/src/core/tracking_copy/mod.rs:285]: Sounds good.

**Review by Fraser999 — APPROVED** [review]

**Review by EdHastingsCasperLabs — APPROVED** [review]

**Pull request merged**